core/src/mender-artifact.c

```diff
@@ -781,18 +793,31 @@ static mender_err_t
 mender_artifact_read_meta_data(mender_artifact_ctx_t *ctx) {
 
     assert(NULL != ctx);
-    size_t index = 0;
 
-    /* Retrieve payload index */
-    if (1 != sscanf(ctx->file.name, "header.tar/headers/%u/meta-data", (unsigned int *)&index)) {
+    /* Retrieve payload index. We expect "header.tar/headers/%u/meta-data" where
+     * %u is the index. Yes sscanf(3) would be nice, but we've experienced
+     * unexplained segmentation faults on some hardware when using it. */
+    const char *const prefix = "header.tar/headers/";
+    if (!mender_utils_strbeginwith(ctx->file.name, prefix)) {
         mender_log_error("Invalid artifact format");
         return MENDER_FAIL;
     }
-    if (index >= ctx->payloads.size) {
+
+    assert(sizeof(size_t) >= sizeof(unsigned long));
+    const char *start_ptr = ctx->file.name + strlen(prefix);
+    char       *end_ptr;
+    errno = 0; /* to distinguish between success/failure */
+
+    const size_t index = strtoul(start_ptr, &end_ptr, 10);
```

Comment:
Not sure if there is any point in preferring `strtoul()` over `strtoull()` here when we have that `assert(sizeof(size_t) >= sizeof(unsigned long long))` and we thus require a big enough `size_t`.